### DISCUSSION TITLE
Is EfficientAD's argument `pad_maps` in the paper?

### DISCUSSION THREAD

**Discussion by jpcbertoldo** [open-edge-platform/anomalib#1368] [Q&A]
The argument 

https://github.com/openvinotoolkit/anomalib/blob/c97762c8781aacfc31d100ba862aa730a3262f7a/src/anomalib/models/efficient_ad/torch_model.py#L234

in the torch model

https://github.com/openvinotoolkit/anomalib/blob/c97762c8781aacfc31d100ba862aa730a3262f7a/src/anomalib/models/efficient_ad/torch_model.py#L214

is used at inference time (`else` is `not self.training`)

https://github.com/openvinotoolkit/anomalib/blob/c97762c8781aacfc31d100ba862aa730a3262f7a/src/anomalib/models/efficient_ad/torch_model.py#L334-L347

before the bilinear interpolation. 

I could not understand where is operation is referenced in the paper. 

Anyone could clarify?  @nelson1425 ?

**Comment by abc-125** [comment]
In the paper in the Appendix E:

> In latency-critical applications, padding in the PDN architecture of EfficientAD can be disabled. This speeds
> up the forward pass of the PDN architecture by 80 μs without impairing the detection of anomalies. We time EfficientAD
> without padding and therefore report the anomaly detection results for this setting in the experimental results of this paper.

So you can disable the padding in PDN like this, using parameter `padding` :
```
pad_mult = 1 if padding else 0
self.conv1 = nn.Conv2d(3, 256, kernel_size=4, stride=1, padding=3 * pad_mult)
```

But if you do this, you need to add padding to the output of the model before resizing it, because it would be smaller than the original image (for 256x256 it would be 56x56 instead of 64x64). You can also check this [answer](https://github.com/nelson1425/EfficientAD/issues/12).

**Reply by jpcbertoldo** [reply]
> In the paper in the Appendix E:
> 
> > In latency-critical applications, padding in the PDN architecture of EfficientAD can be disabled. This speeds
> > up the forward pass of the PDN architecture by 80 μs without impairing the detection of anomalies. We time EfficientAD
> > without padding and therefore report the anomaly detection results for this setting in the experimental results of this paper.
> 
> So you can disable the padding in PDN like this, using parameter `padding` :
> 
> ```
> pad_mult = 1 if padding else 0
> self.conv1 = nn.Conv2d(3, 256, kernel_size=4, stride=1, padding=3 * pad_mult)
> ```

right; but they don't talk about this zero paddng before the upscale and yet the results are supposedly not impacted

> But if you do this, you need to add padding to the output of the model before resizing it, because it would be smaller than the original image (for 256x256 it would be 56x56 instead of 64x64). You can also check this [answer](https://github.com/nelson1425/EfficientAD/issues/12).

sure, it will be 56x56 with `padding=False`, but the bilinear interpolation will make the output 256x256 not matter its input (56 or 64)

in fact, this `pad_maps=True` would create an artifact of 0-only border in the score maps: 

![image](https://github.com/openvinotoolkit/anomalib/assets/24547377/f0b0c31a-2796-488d-9ac8-0581c178411d)


i replied with more detail in https://github.com/nelson1425/EfficientAD/issues/12

**Reply by abc-125** [reply]
I did a quick overlay of this output to an original image, this way it should be more obvious what I am talking about.
With `F.pad(map, (4, 4, 4, 4))`, anomaly map fits the actual image:
![with_padding](https://github.com/openvinotoolkit/anomalib/assets/63813435/6896af7b-add3-413f-a219-ba2e1e9a5a26)
Without it, the anomaly map is cropped closer than the original image and the bottle on it is bigger than the bottle on the image (you can see it from the green glow around the original bottle):
![without_padding](https://github.com/openvinotoolkit/anomalib/assets/63813435/4b9e8b6f-ed45-4556-abce-8df739f78b01)

**Reply by abc-125** [reply]
I also checked nelson1425's implementation (with this padding) on the VisA dataset some time ago, and the results are the same as the paper.

**Reply by jpcbertoldo** [reply]
Hmm, ok I see what you mean.

In this case, it's better to automatically add this post-network paddfing when `padding=False` to avoid confusion/user mistake, no?

**Reply by abc-125** [reply]
> In this case, it's better to automatically add this post-network paddfing when padding=False to avoid confusion/user mistake, no?

Sure, this is a good idea. It can also be worth it to rename `padding` to `padding_conv` to make it more clear which padding it is.

**Reply by jpcbertoldo** [reply]
I added it in 
https://github.com/openvinotoolkit/anomalib/pull/1376
